README.md
# Unicorn HAT Mini
A NodeJS driver for the [Unicorn HAT Mini](https://shop.pimoroni.com/products/unicorn-hat-mini) from [Pimoroni](https://www.pimoroni.com). This software is based on Pimoroni's [official driver and examples](https://github.com/pimoroni/unicornhatmini-python) written in Python.

The device is a 17x7 RGB LED matrix with 4 buttons.

# Installing 
``` bash
npm i unicorn-hat-mini
```

# Demos
Cloning this repository will bring with demos that are not part of the npm package. Sudo must be used to access the GPIO when calling the demo. Ctrl-C can be used to exit the demos.
```bash
# random colors on each pixel every 0.5 seconds
sudo node ./demo/index.js
# moving rainbow across the display, direct port from Python example
sudo node ./demo/rainbow.js
```

## Operating notes
This driver makes use of [node-rpio](https://github.com/jperkin/node-rpio) in the ```gpiomem: false``` mode which requires you to run this as root (```sudo```). Additional troubleshooting and rationale can be found in the node-rpio [README](https://github.com/jperkin/node-rpio#important-system-requirements).

# Usage
``` javascript
const UHM = require('unicorn-hat-mini');
const uhm = new UHM();

uhm.brightness = 0.1;
uhm.setAll(255, 255, 255);
uhm.show();
```
## Combining with other IO functions
To allow control of the IO not used by the HAT, ```UHM.rpio``` is available to set up additional IO and other functionality that the node-rpio library provides.


# API
## UHM(options)
UHM() is a class and must be called with ```new```. It returns an instance of the driver ```uhm```.
``` javascript
const uhm = UHM({brightness: 0.5});
```

|Option|Default|Description|
|---|---|---|
|options.brightness|0.2|Global Brightness of display 0.0-1.0|
|options.enableButtons|true|Use the internal button handler and event emitter. Set to false if you prefer to set up your own access to the HAT's buttons|
|options.exitProcess|true|The driver cleans up the open IO ports when ```SIGINT```, ```SIGTERM```, ```SIGUSR2``` (for compatability with [nodemon](https://github.com/remy/nodemon))

## UHM.ROWS = 7
Constant number of rows on HAT device.

## UHM.COLS = 17
Constant number of columns on HAT device.

## UHM.rpio
Instance of node-rpio provided for allowing configuration of additional IO not used by the HAT.

## uhm.setPixel(row, col, r, g, b)
Set the color of the pixel at (row, col). Must call ```uhm.show()``` for changes to be displayed on the HAT.
``` javascript
// set pixel at row 3 column 5 to purple
uhm.setPixel(3,5,255,0,255);
// Optional variant:
uhm.setPixel(3,5,[255,0,255]);
```

|Parameter|Range|Description|
|---|---|---|
|row|0-3|Vertical position of selected pixel|
|col|0-16|Horizontal position of selected pixel|
|r|0-255|Red value of pixel. Automatically converted to the HAT limits of 0-63. Optional variant: r can be an array of color values ```[r, g, b]```.|
|g|0-255|Green value of pixel. Automatically converted to the HAT limits of 0-63.|
|b|0-255|Blue value of pixel. Automatically converted to the HAT limits of 0-63.|

## uhm.getPixel(row, col)
Get the color of the pixel at (row, col).
``` javascript
// set pixel at row 3 column 5 to purple
uhm.setPixel(3,5,255,0,255);
const pixel = uhm.getPixel(3,5,255,0,255);
// pixel = [255, 0, 255]
```

|Parameter|Range|Description|
|---|---|---|
|row|0-3|Vertical position of selected pixel|
|col|0-16|Horizontal position of selected pixel|

## uhm.setAll(r, g, b)
Set the color of all pixels. Must call ```uhm.show()``` for changes to be displayed on the HAT.
``` javascript
// set all pixels to yellow
uhm.setAll(255,255,0);
// Optional variant:
uhm.setPixel([255,255,0]);
```
|Parameter|Range|Description|
|---|---|---|
|r|0-255|Red value of pixel. Automatically converted to the HAT limits of 0-63. Optional variant: r can be an array of color values ```[r, g, b]```.|
|g|0-255|Green value of pixel. Automatically converted to the HAT limits of 0-63.|
|b|0-255|Blue value of pixel. Automatically converted to the HAT limits of 0-63.|

## uhm.getAll()
Return all pixels as an array in the format: [row][col] = [r, g, b] format
``` javascript
const all = uhm.getAll();
// all = [[[r, g, b], [r, g, b], ...], [[r, g, b], ...], ...]
```

## uhm.show()
Update the display with the latest display buffer as modified by ```uhm.setPixel()``` and ```uhm.setAll()```.
``` javascript
uhm.show();
```

## uhm.clear()
Set all pixels to black/off.
``` javascript
uhm.clear();
```

## uhm.brightness
Set or read the current global brightness of the display 0.0 - 1.0.
``` javascript
// set brightness to 50%
uhm.brightness = 0.5;
// read current brightness
console.log(uhm.brightness)
// displays 0.5
```

## uhm.buttonA, uhm.buttonB, uhm.buttonX, uhm.buttonY
``` javascript
// read state of button A
const buttonState = uhm.buttonA;
```
Returns the state of any of the 4 buttons. The value is not read by accessing this property, instead the driver polls the 4 buttons and produces events (see ```uhm.on```) for change of button states. The most recent state of each button is recorded internally during each event and this most-recent value is returned by these properties. If ```options.enableButtons``` is ```false``` returns undefined.

## uhm.on(event, handler)
A set of events triggered by pressing and releasing buttons. Events can be triggered in the pressed state:
```javascript
event = 
'button-a-pressed'
'button-b-pressed'
'button-x-pressed'
'button-y-pressed'
```
In the released state:
```javascript
event =
'button-a-released'
'button-b-released'
'button-x-released'
'button-y-released'
```
On both state changes:
```javascript
event =
'button-a'
'button-b'
'button-x'
'button-y'
```
Or for any button state change:
```javascript
event =
'button'
```

The handler takes the same callback signature for all button states
```javascript
handler = (button, state) => {
	// button = 'a', 'b', 'x', 'y'
	// state = true for pressed, false for released
}
```
A simple example
```javascript
// displays "button b is true" or "button b is false"
// as the button is pressed and released
uhm.on('button-b', (button, state) => {
	console.log(`Button ${button} is ${state}`);
});
```

# Debugging
Debugging presents some challenges because the node application must be run as sudo. The solution that I have come up with uses [VSCode](https://code.visualstudio.com) and a separately launched [nodemon](https://nodemon.io/) process. Saving a file will trigger nodemon to restart the process and the debugger will need to be connected again.
```json
launch.json
{
	"version": "0.2.0",
	"configurations": [
		{
			"name": "Attach",
			"port": 9229,
			"request": "attach",
			"skipFiles": [
				"<node_internals>/**",
				"**/node_modules/**"
			],
			"type": "node",
			"address": "127.0.0.1",
			"continueOnAttach": true,	// This causes the paused process to continue upon attaching the debugger. It saves a step of having to press F5 (continue) when starting the debugger. It creates the "feel" of launching the process directly inside VSCode
			"localRoot": "${workspaceFolder}",
			"remoteRoot": "/home/ubuntu/projects/unicorn-hat-mini"
		}
	]
}
```

```bash
# run nodemon
# inspect-brk causes the debugger to pause immediately upon entering the program
sudo ./node_modules/nodemon/bin/nodemon.js --inspect-brk demo/rainbow.js
```

CHANGELOG.md
New features and breaking changes are documented here. Corrections to the code are only documented as commits.

# v1.0.0
- Initial release

# v1.1.0
- Add uhm.getPixel()

# v1.2.0
- Add uhm.getAll()

src/lut.js
/* eslint-disable max-len */

// wire to memory mapping for each pixel
// stored separately here to keep the width of the main code file smaller
module.exports = [[139, 138, 137], [223, 222, 221], [167, 166, 165], [195, 194, 193], [111, 110, 109], [55, 54, 53], [83, 82, 81], [136, 135, 134], [220, 219, 218], [164, 163, 162], [192, 191, 190], [108, 107, 106], [52, 51, 50], [80, 79, 78], [113, 115, 114], [197, 199, 198], [141, 143, 142], [169, 171, 170], [85, 87, 86], [29, 31, 30], [57, 59, 58], [116, 118, 117], [200, 202, 201], [144, 146, 145], [172, 174, 173], [88, 90, 89], [32, 34, 33], [60, 62, 61], [119, 121, 120], [203, 205, 204], [147, 149, 148], [175, 177, 176], [91, 93, 92], [35, 37, 36], [63, 65, 64], [122, 124, 123], [206, 208, 207], [150, 152, 151], [178, 180, 179], [94, 96, 95], [38, 40, 39], [66, 68, 67], [125, 127, 126], [209, 211, 210], [153, 155, 154], [181, 183, 182], [97, 99, 98], [41, 43, 42], [69, 71, 70], [128, 130, 129], [212, 214, 213], [156, 158, 157], [184, 186, 185], [100, 102, 101], [44, 46, 45], [72, 74, 73], [131, 133, 132], [215, 217, 216], [159, 161, 160], [187, 189, 188], [103, 105, 104], [47, 49, 48], [75, 77, 76], [363, 362, 361], [447, 446, 445], [391, 390, 389], [419, 418, 417], [335, 334, 333], [279, 278, 277], [307, 306, 305], [360, 359, 358], [444, 443, 442], [388, 387, 386], [416, 415, 414], [332, 331, 330], [276, 275, 274], [304, 303, 302], [337, 339, 338], [421, 423, 422], [365, 367, 366], [393, 395, 394], [309, 311, 310], [253, 255, 254], [281, 283, 282], [340, 342, 341], [424, 426, 425], [368, 370, 369], [396, 398, 397], [312, 314, 313], [256, 258, 257], [284, 286, 285], [343, 345, 344], [427, 429, 428], [371, 373, 372], [399, 401, 400], [315, 317, 316], [259, 261, 260], [287, 289, 288], [346, 348, 347], [430, 432, 431], [374, 376, 375], [402, 404, 403], [318, 320, 319], [262, 264, 263], [290, 292, 291], [349, 351, 350], [433, 435, 434], [377, 379, 378], [405, 407, 406], [321, 323, 322], [265, 267, 266], [293, 295, 294], [352, 354, 353], [436, 438, 437], [380, 382, 381], [408, 410, 409], [324, 326, 325], [268, 270, 269], [296, 298, 297]];


demo/rainbow.js
// launch the process on the server using sudo:
// sudo ./node_modules/nodemon/bin/nodemon.js --inspect-brk=0.0.0.0:9229 test/index.js

const UHM = require('../src/index');
const HSVtoRGB = require('./hsvtorgb');

const uhm = new UHM();

uhm.brightness = 0.1;
uhm.setAll(255, 255, 255);
uhm.show();

let step = 0;
const { ROWS, COLS } = UHM;

setInterval(() => {
	step += 1;
	for (let r = 0; r < ROWS; r += 1) {
		for (let c = 0; c < COLS; c += 1) {
			const dx = Math.sin(step / COLS + 20) * COLS + ROWS;
			const dy = Math.cos(step / ROWS) * ROWS + ROWS;
			const sc = Math.cos(step / ROWS) * ROWS + COLS;
			const hue = Math.sqrt((c - dx) ** 2 + (r - dy) ** 2) / sc;
			const rgb = HSVtoRGB(hue, 1, 1);
			uhm.setPixel(r, c, rgb);
		}
	}
	uhm.show();
}, 50);

uhm.on('button', (button, state) => console.log(button, state));


src/index.js
// access a Pimoroni Unicorn HAT Mini over the i2c bus
// the HAT makes use of two HT16D35A chips

const rpio = require('rpio');

const EventEmitter = require('events');

const DEFAULT_OPTIONS = {
	exitProcess: true,
	brightness: 0.2,
	enableButtons: true,
};

/* eslint-disable no-unused-vars */
const CMD_SOFT_RESET = 0xCC;
const CMD_GLOBAL_BRIGHTNESS = 0x37;
const CMD_COM_PIN_CTRL = 0x41;
const CMD_ROW_PIN_CTRL = 0x42;
const CMD_WRITE_DISPLAY = 0x80;
const CMD_READ_DISPLAY = 0x81;
const CMD_SYSTEM_CTRL = 0x35;
const CMD_SCROLL_CTRL = 0x20;

const COLS = 17;
const ROWS = 7;
/* eslint-enable no-unused-vars */

const BUTTON_A = 5;
const BUTTON_B = 6;
const BUTTON_X = 16;
const BUTTON_Y = 24;

const LUT = require('./lut');

// display is handled by two chips at addresses:
// 0 = left, 1 = right
const DEVICES = [
	{
		address: 0,
		cs: 0,
		offset: 0,
	}, {
		address: 1,
		cs: 1,
		offset: 28 * 8,
	},
];

const sendDevice = (dev, data) => {
	const txBuffer = Buffer.from(data);
	rpio.spiChipSelect(dev.cs);
	rpio.spiWrite(txBuffer, txBuffer.length);
};

// convenience function to run tasks on both HT16D35A chips
const bothDevices = (callback) => {
	DEVICES.forEach((dev) => callback(dev));
};

class UHM extends EventEmitter {
	#buttons = {
		a: undefined,
		b: undefined,
		x: undefined,
		y: undefined,
	};

	// cleanup monitoring
	#exited = false;

	// internal display management
	#buffer = [];

	#display = [];

	constructor(options) {
		super();
		this.opts = { ...DEFAULT_OPTIONS, options };

		// fill buffer with all zeros initially
		for (let i = 0; i < 28 * 8 * 2; i += 1) {
			this.#buffer.push(0);
		}

		// set display to all black
		for (let i = 0; i < ROWS * COLS; i += 1) {
			this.#display.push([0, 0, 0]);
		}

		// configure rpio
		rpio.init({
			mapping: 'gpio',
			// required for spi
			gpiomem: false,
			// define our own shutdown method
			close_on_exit: false,
		});

		// configure the spi channel
		rpio.spiBegin();
		rpio.i2cSetClockDivider(250); // 1 MHz

		// cleanup on exit
		const exit = (type) => {
			// only exit once
			if (!this.#exited) {
				bothDevices((dev) => {
					sendDevice(dev, [CMD_COM_PIN_CTRL, 0x00]);
					sendDevice(dev, [CMD_ROW_PIN_CTRL, 0x00, 0x00, 0x00, 0x00]);
					sendDevice(dev, [CMD_SYSTEM_CTRL, 0x00]);
				});
				rpio.spiEnd();
				rpio.exit();
				this.#exited = true;
			}
			if (this.opts.exitProcess && type !== 'exit') {
				process.exit();
			}
		};

		process.on('SIGINT', exit);
		process.on('SIGTERM', exit);
		process.on('exit', () => exit('exit'));
		// for use with nodemon
		process.on('SIGUSR2', exit);

		// configure each device
		bothDevices((dev) => {
			sendDevice(dev, [CMD_SOFT_RESET]);
			sendDevice(dev, [CMD_GLOBAL_BRIGHTNESS, 0x01]);
			sendDevice(dev, [CMD_SCROLL_CTRL, 0x00]);
			sendDevice(dev, [CMD_SYSTEM_CTRL, 0x00]);
			sendDevice(dev, [CMD_WRITE_DISPLAY, 0x00, ...this.#buffer.slice(dev.offset, dev.offset + 28 * 8)]);
			sendDevice(dev, [CMD_COM_PIN_CTRL, 0xff]);
			sendDevice(dev, [CMD_ROW_PIN_CTRL, 0xff, 0xff, 0xff, 0xff]);
			sendDevice(dev, [CMD_SYSTEM_CTRL, 0x03]);
		});

		// initial brightness
		this.brightness = this.opts.brightness;

		// set up buttons
		if (this.opts.enableButtons) {
			// configure input
			rpio.open(BUTTON_A, rpio.INPUT, rpio.PULL_UP);
			rpio.open(BUTTON_B, rpio.INPUT, rpio.PULL_UP);
			rpio.open(BUTTON_X, rpio.INPUT, rpio.PULL_UP);
			rpio.open(BUTTON_Y, rpio.INPUT, rpio.PULL_UP);

			// read initial state
			this.#buttons.a = rpio.read(BUTTON_A);
			this.#buttons.b = rpio.read(BUTTON_B);
			this.#buttons.x = rpio.read(BUTTON_X);
			this.#buttons.y = rpio.read(BUTTON_Y);

			// set up polling
			rpio.poll(BUTTON_A, () => this.#buttonHandler(BUTTON_A, rpio.POLL_BOTH));
			rpio.poll(BUTTON_B, () => this.#buttonHandler(BUTTON_B, rpio.POLL_BOTH));
			rpio.poll(BUTTON_X, () => this.#buttonHandler(BUTTON_X, rpio.POLL_BOTH));
			rpio.poll(BUTTON_Y, () => this.#buttonHandler(BUTTON_Y, rpio.POLL_BOTH));
		}
	}

	// handle a button press
	#buttonHandler(button) {
		// button is pressed when in the low state
		const state = rpio.read(button) === rpio.LOW;
		let identifier;

		// check button and set current state
		switch (button) {
		case BUTTON_A:
			identifier = 'a';
			this.#buttons.a = state;
			break;
		case BUTTON_B:
			identifier = 'b';
			this.#buttons.b = state;
			break;
		case BUTTON_X:
			identifier = 'x';
			this.#buttons.x = state;
			break;
		case BUTTON_Y:
			identifier = 'y';
			this.#buttons.y = state;
			break;
		default:
		}

		// catch misconfiguration
		if (!identifier) return;

		// dispatch events
		// -on and -off are only triggered when the switches to these states
		// 'button-' event is emitted for any change of state
		if (state) {
			this.emit(`button-${identifier}-pressed`, identifier, true);
		} else {
			this.emit(`button-${identifier}-released`, identifier, false);
		}
		this.emit(`button-${identifier}`, identifier, state);
		this.emit('button', identifier, state);
	}

	get buttonA() {
		return this.#buttons.a;
	}

	get buttonB() {
		return this.#buttons.b;
	}

	get buttonX() {
		return this.#buttons.x;
	}

	get buttonY() {
		return this.#buttons.y;
	}

	setPixel(row, col, r, g, b) {
		// allow r to be an array of colors
		if (Array.isArray(r)) {
			/* eslint-disable prefer-destructuring, no-param-reassign */
			g = r[1];
			b = r[2];
			r = r[0];
			/* eslint-enable prefer-destructuring, no-param-reassign */
		}

		// determine the offset
		const offset = (row * COLS) + col;
		this.#display[offset] = [r, g, b];
	}

	getPixel(row, col) {
		const offset = (row * COLS) + col;
		const shifted = this.#display[offset];
		return [shifted[0], shifted[1], shifted[2]];
	}

	setAll(r, g, b) {
		// allow r to be an array of colors
		/* eslint-disable prefer-destructuring, no-param-reassign */
		if (Array.isArray(r)) {
			g = r[1];
			b = r[2];
			r = r[0];
		}
		/* eslint-enable prefer-destructuring, no-param-reassign */

		for (let i = 0; i <= ROWS * COLS; i += 1) {
			this.#display[i] = [r, g, b];
		}
	}

	// return all pixels in [row][col] = [r, g, b] format
	getAll() {
		const all = [];
		for (let r = 0; r < ROWS; r += 1) {
			const row = [];
			for (let c = 0; c < COLS; c += 1) {
				row.push(this.#display[r * COLS + c]);
			}
			all.push(row);
		}
		return all;
	}

	clear() {
		this.setAll(0, 0, 0);
	}

	get brightness() {
		return this.opts.brightness;
	}

	set brightness(b = 0.3) {
		// b must be between 0 and 1
		const brightness = Math.max(0, Math.min(1, b));
		this.opts.brightness = brightness;
		bothDevices((dev) => {
			sendDevice(dev, [CMD_GLOBAL_BRIGHTNESS, Math.floor(63 * brightness)]);
		});
	}

	show() {
		for (let i = 0; i < ROWS * COLS; i += 1) {
			const [ir, ig, ib] = LUT[i];
			const [r, g, b] = this.#display[i];
			// shift to 0-63
			this.#buffer[ir] = r >> 2;
			this.#buffer[ig] = g >> 2;
			this.#buffer[ib] = b >> 2;
		}
		bothDevices((dev) => {
			sendDevice(dev, [CMD_WRITE_DISPLAY, 0x00, ...this.#buffer.slice(dev.offset, dev.offset + 28 * 8)]);
		});
	}

	static get rpio() {
		return rpio;
	}

	static get ROWS() {
		return ROWS;
	}

	static get COLS() {
		return COLS;
	}
}

module.exports = UHM;
